runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -11658,12 +11659,106 @@ J9::Power::CodeGenerator::inlineDirectCall(TR::Node *node, TR::Register *&result
+            //     cannot determine the final destination address right away, and thus need to keep dest and destOffset separate. In this
+            //     case, the node passed in to setmemoryEvaluator() will have four children: dest, destOffset, len, and byteValue.
+            //
+            // Note that if the default (gencon) GC policy is being used, dataAddr will never need to be loaded no matter
+            // the type of the object at dest, so all of the above cases will be treated like case (1).
+
+            //if offheap is not enabled, we will never need to load dataAddr. So we know for sure that we can calculate the final
+            //destination address right away and only pass in three children to setmemoryEvaluator() (dest, len, byteValue)
+            bool loadDataAddr = false;
+            bool separateDestAndOffset = false;
+
+         #if defined (J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+            if (TR::Compiler->om.isOffHeapAllocationEnabled() && comp->target().is64Bit())
+               {
+               if (dest->getSymbolReference() != NULL)
+                  {
+                  //check dstBaseAddrNode type at compile time
+                  int length;
+                  const char *objTypeSig = dest->getSymbolReference()->getTypeSignature(length);
+
+                  // There are four cases where we cannot be sure of the Object type at compile time:
+                  // 1.) The object's type signature is unavailable/unknown
+                  // 2.) The signature names java/lang/Object
+                  // 3.) The signature names some interface type, wich is a problem because interface types are
+                  //     not checked by the verifier
+                  // 4.) The signature belongs to a parameter of the calling method, which is a problem because
+                  //     parameters can share slots with variables of other, potentially incompatible types
+                  bool objTypeUnknown;
+                  TR_OpaqueClasBlock *objClass = NULL;
+
+                  if (objTypeSig == NULL)
+                     {
+                     objTypeUnknown = true;
+                     }
+                  else
+                     {
+                     objClass = fej9->getClassFromSignature(objTypeSig, length, dest->getSymbolReference()->getOwningMethod(comp));
+                     bool isParameter = dest->getSymbolReference()->getSymbol()->isParm();
+
+                     objTypeUnknown = objClass == NULL || objClass == comp->getObjectClassPointer() ||
+                                      TR::Compiler->cls.isInterfaceClass(comp, objClass) || isParameter;
+                     }
+
+                  //need to generate code to loadDataAddr in cases (2) and (3) (unless the object is known to be NULL)
+                  loadDataAddr = !dest->isNull() && (objTypeUnknown || objTypeSig[0] == '[');
```

Comment:
`!node->isNull()`  is different from  `node->isNonNull()`.  the former doesn't guarantee it is not NULL. I figured you were looking for the latter to use here (?).

Reply:
loadDataAddr seemed not the intention anyway. variable name can mislead the reading, just aggravating the situation. it appeared to me just meaning "we don't know whether we need to load dataAddr" (since you have objTypeUnknown as the positive condition).

Reply:
I believe the `!node->isNull()` condition is being used correctly here. If the object is a `NULL` reference, we can't load `dataAddr` from it. So if we already know that this is the case at compile time (i.e.: `node->isNull()`),  then we can set `loadDataAddr` to false right away. On the other hand, even if we know at compile time that the object is NOT a `NULL` reference (i.e.: `node->isNonNull()`), we don't know whether or not we will need to load `dataAddr`, so it doesn't have any effect on this condition. 

That being said, I can certainly see how the name `loadDataAddr` makes this confusion. Would something like `generatePathToLoadDataAddr` be more clear? Alternatively I can change the logic for setting this condition so that it matches the name better (i.e.: `loadDataAddr` evaluates to true ONLY when we know for sure that we need to load dataAddr).

Reply:
`mightNeedDataAddr`  or something like that